util/extsort/disk_sorter.go

```diff
@@ -131,29 +900,363 @@ func (d *DiskSorter) Sort(ctx context.Context) error {
+			endKey = file.endKey
+		}
+	}
+	// If there is kv stats, return a single compaction for all files.
+	if len(buckets) == 0 {
+		overlapFiles := slices.Clone(files)
+		slices.SortFunc(overlapFiles, func(a, b *fileMetadata) bool {
+			return bytes.Compare(a.startKey, b.startKey) < 0
+		})
+		return []*compaction{{
+			startKey:     startKey,
+			endKey:       endKey,
+			overlapFiles: overlapFiles,
+		}}
+	}
+
+	slices.SortFunc(buckets, func(a, b kvStatsBucket) bool {
+		return bytes.Compare(a.UpperBound, b.UpperBound) < 0
+	})
+	// Merge buckets with the same upper bound.
+	n := 0
+	for i := 1; i < len(buckets); i++ {
+		if bytes.Equal(buckets[n].UpperBound, buckets[i].UpperBound) {
+			buckets[n].Size += buckets[i].Size
+		} else {
+			n++
+			buckets[n] = buckets[i]
+		}
+	}
+	buckets = buckets[:n+1]
+
+	var (
+		kvSize      int
+		compactions []*compaction
+	)
+	for i, bucket := range buckets {
+		if i+1 == len(buckets) {
+			compactions = append(compactions, &compaction{
+				startKey: startKey,
+				endKey:   endKey,
+			})
+			break
+		}
+		kvSize += bucket.Size
+		if kvSize >= maxCompactionSize {
+			compactions = append(compactions, &compaction{
+				startKey: startKey,
+				endKey:   bucket.UpperBound,
+			})
+			startKey = bucket.UpperBound
+			kvSize = 0
+		}
+	}
+
+	for _, c := range compactions {
+		for _, file := range files {
+			if !(bytes.Compare(file.endKey, c.startKey) <= 0 ||
+				bytes.Compare(file.startKey, c.endKey) >= 0) {
+				c.overlapFiles = append(c.overlapFiles, file)
+			}
+		}
+		slices.SortFunc(c.overlapFiles, func(a, b *fileMetadata) bool {
+			return bytes.Compare(a.startKey, b.startKey) < 0
+		})
+	}
+	return compactions
+}
+
+func (d *DiskSorter) runCompaction(ctx context.Context, c *compaction) (*fileMetadata, error) {
+	d.opts.Logger.Debug(
+		"run compaction",
+		zap.Binary("startKey", c.startKey),
+		zap.Binary("endKey", c.endKey),
+		zap.Int("fileCount", len(c.overlapFiles)),
+	)
+
+	var merged *fileMetadata
+	onSuccess := func(meta *fileMetadata) { merged = meta }
+
+	fileNum := int(d.idAlloc.Add(1))
+	sw, err := newSSTWriter(d.fs, d.dirname, fileNum, defaultKVStatsBucketSize, onSuccess)
+	if err != nil {
+		return nil, err
+	}
+	swClosed := false
+	defer func() {
+		if !swClosed {
+			_ = sw.Close()
+		}
+	}()
+
+	iter := newMergingIter(c.overlapFiles, d.openIter)
+	defer func() {
 		_ = iter.Close()
-		return errors.Trace(iter.Error())
+	}()
+
+	iterations := 0
+	for iter.Seek(c.startKey); iter.Valid(); iter.Next() {
+		if bytes.Compare(iter.UnsafeKey(), c.endKey) >= 0 {
+			break
+		}
+		iterations++
+		if iterations%1000 == 0 && ctx.Err() != nil {
+			return nil, errors.Trace(ctx.Err())
+		}
+		if err := sw.Set(iter.UnsafeKey(), iter.UnsafeValue()); err != nil {
+			return nil, err
+		}
 	}
-	end := slices.Clone(iter.Key())
-	if err := iter.Close(); err != nil {
-		return errors.Trace(err)
+	if err := iter.Error(); err != nil {
+		return nil, err
+	}
+	err = sw.Close()
```

Comment:
if the process exit after this line and before 
```
					_ = d.fs.Remove(makeFilename(d.fs, d.dirname, file.fileNum))
```
the data is duplicated, and we must rely on the deduplication when Next()

Reply:
That's right.